### ISSUE TITLE
ArrayIndexOutOfBoundsException when reading String from ResultSet

### ISSUE THREAD

**Issue by Nik--** [brettwooldridge/HikariCP#1712]
The following exception is being thrown when reading a specific piece of data from a *ResultSet*.
```
 java.lang.ArrayIndexOutOfBoundsException: Index 52 out of bounds for length 52
	at com.mysql.cj.mysqla.io.Buffer.readInteger(Buffer.java:284)
	at com.mysql.cj.mysqla.result.BinaryBufferRow.getValue(BinaryBufferRow.java:224)
	at com.mysql.cj.jdbc.result.ResultSetImpl.getString(ResultSetImpl.java:954)
	at com.mysql.cj.jdbc.result.ResultSetImpl.getString(ResultSetImpl.java:966)
	at com.zaxxer.hikari.pool.HikariProxyResultSet.getString(HikariProxyResultSet.java)
```

I have observed that the exception does not appear when `dataSource.cacheResultSetMetadata` is turned off, which makes me believe that is some kind of cache issue.

The configuration used is:
```properites
dataSource.cachePrepStmts=true
dataSource.prepStmtCacheSize=250
dataSource.prepStmtCacheSqlLimit=2048
dataSource.useServerPrepStmts=true
dataSource.useLocalSessionState=true
dataSource.rewriteBatchedStatements=true
dataSource.cacheResultSetMetadata=true
dataSource.cacheServerConfiguration=true
dataSource.elideSetAutoCommits=true
dataSource.maintainTimeStats=false
leakDetectionThreshold=600000
```
The driver used is: `mysql-connector-java 6.0.3`.
The database server used is: `mysql  Ver 15.1 Distrib 10.5.8-MariaDB, for Linux (x86_64) using readline 8.0`

The query used is:
```sql
"SELECT int1, string1, string2, text1, text2 FROM tableName"
```
With table:
```
DROP TABLE IF EXISTS `table_name`;
CREATE TABLE `table_name` (
  `int1` smallint(5) NOT NULL DEFAULT 0,
  `string1` varchar(25) NOT NULL,
  `string2` varchar(70) NOT NULL,
  `text1` text DEFAULT NULL,
  `text2` text DEFAULT NULL,
  PRIMARY KEY (`int1`)
) ENGINE=MyISAM DEFAULT CHARSET=latin1;

INSERT INTO `table_name` VALUES (112,'N1','Unassigned','',NULL),(113,'N2','Default','FirstWaitDelay-30000,WaitDelay-20000',NULL);
```

This is an example of the java part which throws the exception:
```java
public static void main(String[] args) throws Exception
	{
		final HikariConfig config = new HikariConfig("hikari.properties");
		config.setDriverClassName(DRIVER);
		config.setJdbcUrl(URL);
		config.setUsername(USERNAME);
		config.setPassword(PASSWORD);
		DataSource ds = new HikariDataSource(config);
		
		try (Connection con = ds.getConnection();
			 PreparedStatement statement = con.prepareStatement("SELECT `int1`, `string1`, `string2`, `text1`, `text2` FROM table_name");
			 ResultSet rset = statement.executeQuery())
		{
			while (rset.next())
			{
				rset.getString("string2");
				rset.getInt("int1");
				rset.getString("string1");
				rset.getString("text2");
				rset.getInt("int1");
				rset.getString("text1");
			}
		}
		catch (SQLException e)
		{
			e.printStackTrace();
		}
	}
```

**Comment by Nik--** [comment]
Updating database driver to `mysql-connector-java 8.0.22` did solve the issue. Though it seems to be quite the rare issue because of dozens of tables it just ocurred on one table and only some of its records did throw it.

**Comment by quaff** [comment]
Why do you think it's related to HikariCP not mysql driver?

**Comment by Nik--** [comment]
Well the issue occurred after moving from C3P0 to HikariCP, therefore it was the first suspect on the line. There were no issues with C3P0, but then again, it most likely did not cache the result set metadata which caused the issue at HikariCP on the first place. Even though I found a solution to the issue, I decided to leave it open for the contributors to decide if some changes are needed, because the error was pretty vague and not giving any hint to problematic driver or whatsoever. Feel free to close this issue if nobody thinks there is need for such enhancements.
